### DISCUSSION TITLE
Create a factory for actions using createAction not working

### DISCUSSION THREAD

**Discussion by ChakirMrabet** [ngrx/platform#3361] [Q&A] [answered]
I want to create a function that returns an action function using the `createAction` function:

```
function actionFactory<T extends object>(description: string) {
  return createAction(
    `[${FEATURE_STATE_NAME}] ${description}`,
    props<T>() as ActionCreatorProps<T> & NotAllowedCheck<T>
  );
}
```

I want to do this to simplify the creation of actions in my project.

I get the following error:

```
 error TS2344: Type 'T' does not satisfy the constraint 'NotAllowedInPropsCheck<T>'.
  Type 'object' is not assignable to type 'NotAllowedInPropsCheck<T>'.

18     props<**T**>() as ActionCreatorProps<T> & NotAllowedCheck<T>
```

Any ideas of what I'm doing wrong?

I'm using NgRx version 13.

Thanks.

**Comment by david-shortman** [comment, marked as answer]
`T` must be guaranteed to satisfy the constraint `NotAllowedInPropsCheck`. This type guard excludes objects with a `type` property, empty objects, and primitives.

I am not aware of a way to type this such that a generic declared in `actionFactory` is guaranteed to pass this check. It only works with concrete object types.

However, if you take in a "props config" as a parameter, then you could make a factory:

```ts
const FEATURE_STATE_NAME = 'Some Feature';

function actionFactory<D extends string, P extends object>(description: D, /* include this param 👉 */ config: ActionCreatorProps<P> & NotAllowedCheck<P>) {
  return createAction(`[${FEATURE_STATE_NAME}] ${description}`, config);
}

// 👇 then pass a props instance as an argument
const ExampleAction = actionFactory('Example Action', props<{ some: string }>());

// the action works as expected
const exampleInstance: { some: string; } & TypedAction<'[Some Feature] Example Action'> = ExampleAction({ some: 'value' });
```

**Reply by ChakirMrabet** [reply]
Thanks!! That's a clever way of doing it!

I've just had to refactor it a little bit to allow actions with no props:

```typescript
function actionFactory<D extends string, P extends object>(
  description: D,
  props?: ActionCreatorProps<P> & NotAllowedCheck<P>
) {
  return props
    ? createAction(`[${FEATURE_STATE_NAME}] ${description}`, props)
    : createAction(`[${FEATURE_STATE_NAME}] ${description}`);
}
```

Without the check to see if props is defined, I get an error from `createAction`. It seems it doesn't allow the `props` argument to be `undefined` or `null`.
